### PR TITLE
Refactor refresh mechanism

### PR DESCRIPTION
**Pull request by rotemavni**
Relates to issue GoogleCloudPlatform#744

**This PR includes:**

Instead of importing each service sequentially, do the following:

1. Init resources for all services using threads.
2. Aggregate all resources and shuffle them to a random order
3. Call Refresh for all the resources together.

To implement this logic I've also added:

1. RetryCount and RetrySleepMs to import flags, this helps customize these parameters for the retry if Refresh failed.
2. ProvidersMapping - a struct that helps keep track of the relationships between provider, service and resources. There is a provider per service, and we want to be able to trace back the refreshed resources to the service and provider they belong to.

### CODE REVIEW

**Review by meshuga — CHANGES_REQUESTED** [review]
Thanks for the PR, the changes do speed up the logic significantly and the change doesn't much affect the resulting files.

Terraformer is used in various ways, one of the way is to generate code out of the existing infrastructure, in a form of inverted IoC, where resources are fetched from an infrastructure and Terraform resource files are being generated and stored. This approach allows an easy tracking of changes of the infrastructure changes by generating diffs of the resource files with e.g. Git. Your PR introduces randomization of resources in func `ShuffleResources` in `ProvidersMapping`, which prevents such diff to be created. There was an effort being done in e.g. #469 to ensure repeatable files on each run and it is something that I want to keep having. Can you ensure the **order of resources** in resulting file, that is either remove resource randomisation (not sure if it's really helpful) or add a logic to sort resulting resources by a tuple of its resource type and identifier.

Another thing that I am worried is that you parallelized service execution with an assumption various services will be handled by different servers. That's not always the case. Terraformer does support single servers for e.g. Kubernetes provider. In case of Keycloak it's not a case as there's only one service. On the other hand, in case of a single server overload, a user can play with retries to ensure calls will be executed successfully.

File: cmd/import.go [hunk 1]
```diff
@@ -112,64 +136,83 @@ func Import(provider terraformutils.ProviderGenerator, options ImportOptions, ar
...
-		resources, err := buildServiceResources(service, provider, options, providerWrapper)
+		serviceProvider := providersMapping.AddServiceToProvider(service)
+		err := serviceProvider.Init(args)
 		if err != nil {
-			log.Println(err)
-			continue
+			return err
 		}
-		plan.ImportedResource[service] = append(plan.ImportedResource[service], resources...)
+		go initServiceResourcesWorker(service, serviceProvider, options, providerWrapper, &wg, failedServices)
```
Comment (sergeylanzman) [inline cmd/import.go:160]: I think we can use not parallel option here, as today.

File: terraformutils/providerwrapper/provider.go [hunk 2]
```diff
@@ -56,12 +56,23 @@ type ProviderWrapper struct {
...
+	retrySleepMs int
 }
 
-func NewProviderWrapper(providerName string, providerConfig cty.Value, verbose bool) (*ProviderWrapper, error) {
+func NewProviderWrapper(providerName string, providerConfig cty.Value, verbose bool, retryOptions ...int) (*ProviderWrapper, error) {
 	p := &ProviderWrapper{}
 	p.providerName = providerName
 	p.config = providerConfig
+
+	if len(retryOptions) == 2 {
```
Comment (sergeylanzman) [inline terraformutils/providerwrapper/provider.go:68]: Why it's not can be separate? If user want pass only retryCount?

**Comment by sergeylanzman** [comment]
@meshuga 
+1 not use parallel request as today, most of resources it's single huge request or paginations

About sort resulting
Today we have issue with sorting resources in final tf files. It's because we use golang map and it's not sorted by design. ShuffleResources not changed this. because it's saved results to same map and map not sorted.
it's not matter if you refresh VPC1,SG1,VPC3,SG4 anyway it's saved to different not sorted maps
https://github.com/GoogleCloudPlatform/terraformer/blob/785be17c812c39e2462333371f42a45d5e5976bd/terraformutils/hcl.go#L237-L277

**Comment by nimrodkor** [comment]
Hi @sergeylanzman @meshuga, thanks for the feedback!

Just wanted to make sure I understood all of your comments
1. Sorting - per @sergeylanzman ’s comment, order isn’t promised as it is today so not sure we have something to do there. If I misunderstood, we can add a flag and let the user decide whether he wants order or speed.
2. Parallel init - you both noted it, we’ll implement it.
3. Retry options - will de as well

**Comment by rotemavni** [comment]
Hi @sergeylanzman @meshuga,
I've added the discussed changes. Please let me know what you think.
We appreciate your review, thanks!

File: terraformutils/utils.go [hunk 3]
```diff
@@ -65,40 +65,86 @@ func PrintTfState(resources []Resource) ([]byte, error) {
...
+	input := make(chan *Resource, len(resources))
 	var wg sync.WaitGroup
 	poolSize := 15
-	if slowProcessingRequired(resources) {
-		poolSize = 1
+	for i := range resources {
+		wg.Add(1)
+		input <- resources[i]
 	}
+	close(input)
```
Comment (sergeylanzman) [inline terraformutils/utils.go:77]: why it's here?

Comment (rotemavni) [reply terraformutils/utils.go:77]: @sergeylanzman Because we're done adding data to the channel at that point.

**Pull request merged**